Add update subcommand

jonasfreimuth opened the pull request:
General workflow:

* Add handles whose people regularly participate in the lab meeting (`/sudo update add $HANDLES`)
* Let the bot choose any number of them at random, weighted by how long it has been since their last presentation (`/sudo update choose [$N]`)
* If one of them actually presents, confirm this to update their last update date to the current time (`/sudo update confirm $HANDLES`)

jonasfreimuth committed:
Add time format now

  Should lead to posting of the reminder right away. There are
  some uses of this beyond debugging, and with some time I might even
  think of one.

jonasfreimuth committed:
Display error if message posting fails

  This way I noticed I am missing gnutls, I'm sure this could help
  others in similar situations.

jonasfreimuth committed:
Factor subcommand out from controller

  Should make it easier to add new subcommands down the line.

jonasfreimuth committed:
Change hook url to be variable

  Requires additional parameters for the app, but that seems alright for
  better flexibility.

jonasfreimuth committed:
Add update table to db

  Will contain user handles and last time of participation of the user
  in the update round.

jonasfreimuth committed:
subcommand-make: Add handling of unexpected thing

jonasfreimuth committed:
Switch subcommand-make to take token list

  Saves some repeating of the "make" keyword.

jonasfreimuth committed:
Switch subcommand-remind to take token list

  Saves some repeating of the "remind" keyword. Made some changes to
  downstream procedures necessary.

jonasfreimuth committed:
Add proc handle-exists?

  Checks if a handle is present in the current db.

jonasfreimuth committed:
Add proc add-handles

  Adds handles to the current db, recording the time and returning a
  string describing the result for each handle.

jonasfreimuth committed:
Add proc remove-handles

  Removes handles from the current db, recording the time and returning
  a string describing the result for each handle.

jonasfreimuth committed:
Add proc list-handles

  Return a markdown formatted list of all handles and
  their last presentation date.

jonasfreimuth committed:
Add proc choose-candidate

  Return a markdown formatted list of all handles and
  their last presentation date.

jonasfreimuth committed:
Add proc for update subcommand

  This subcommand should be used to choose candidates for a presentation
  of some kind. Candidates that have not presented in a while are
  supposed to have a higher likelihood to be chosen.

jonasfreimuth committed:
Add proc confirm-candidate

  Updates the time of last update for all given handles to the current
  time and returns a string informing about the success / fail for each.

jonasfreimuth committed:
Enable update subcommand

jonasfreimuth committed:
Add help text for remind subcommand

jonasfreimuth committed:
Add help text for make subcommand

jonasfreimuth committed:
Add help text for update subcommand

jonasfreimuth committed:
Add proper overall help text

jonasfreimuth committed:
Remove old help-text variable

  No longer needed.

jonasfreimuth committed:
Move reminder prefix of message to reminder proc

  Makes post-message proc more versatile.

jonasfreimuth committed:
Move error handling into post-message proc

  Should make this procedure more resilient. Not sure if this is
  entirely appropriate.

jonasfreimuth committed:
Make update procs output consistent

  Now they all output a single string, with information generally
  separated with newlines.

jonasfreimuth committed:
Add parenthesis to choose complaint string

  It's not the main message after all.

jonasfreimuth committed:
Change choose subcommand to post to town-square

  The private webhook part is kept as a confirmation, if the command is
  invoked from other channels.

jonasfreimuth committed:
Add proc weighted-random

  Will be used for candidate selection.

jonasfreimuth committed:
Add proc weighted-sample

  Will be used for candidate selection.

jonasfreimuth committed:
Update choose-candiate to use weighted random

jonasfreimuth committed:
Add confirmation string to choose-candidates

jonasfreimuth committed:
Add general workflow section to update-help

jonasfreimuth commented on mattermost.scm:
```diff
@@ -604,6 +604,44 @@ a colon-separated pair of hours and minutes."
+        `(("text" . ,(string-join (remove-handles handles)
+                                   ", ")))))
+    (("choose" . n-list)
+      (let ((choose-string (if (null? n-list)
```
This is probably unsafe, there are no checks that this is actually a number. If it isn't the result would probably be an unhandled error.

rekado replied:
is `n-list` supposed to be a number?

I'd do `("choose" n)` instead of `("choose" . n-list)` because that guarantees that `n` is an atom.

You can also tell `match` to run arbitrary predicates:

    ("choose" (? string->number n))

This would match only on the list where the second element is a string that can be converted to a number (i.e. where `string->number` returns something other than `#false`).

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+(define (weighted-random weights)
+  ;; TODO Ensure no zeros in weights.
+  (let* ((weights-sum (apply + weights))
+         (thresh (random weights-sum)))
```
note that `random` requires initialization or else it will return the same sequence of "random" numbers every time the program is restarted.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+      (if (or (> weights-accum thresh)
+              (>= loop-idx (length weights)))
+          loop-idx
+          (loop (+ loop-idx 1) (+ weights-accum (list-ref weights (+ loop-idx 1))))))))
```
I think I'd use `fold` here instead of manually carrying `weights-accum` from one iteration to the next.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+          (assoc-remove! x (car choice-pair))
+          (+ n 1)
+          n-max
+          (append choices (list (car choice-pair)))))))
```
the more idiomatic way is to `(cons choice-pair choices)` and reverse the list if necessary.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+                         (- (time-second (current-time time-utc)) date))
+                        (map
+                         (lambda (vector) (vector-ref vector 1))
+                         results))))
```
I'd probably use `match` on `results` here to extract all first vector elements.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+         (map (lambda (entry)
+                (let ((handle (vector-ref entry 0))
+                     (last-update (vector-ref entry 1)))
+                  (format
```
You can use `match-lambda` to destructure `entry` instead of using `vector-ref`.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+
+(define (confirm-candidates handles)
+  (string-join
+   (map
```
This is a little weird.  You're evaluating the lambda only for side effects; the only reason you use `map` is because you collect strings to be output at a later time.  I feel that this conflates two things that I'd rather see separated.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+                   handle
+                   (readable-date-from-seconds last-update)
+                   handle)))
+          (weighted-sample results weights (string->number n)))
```
I'd change choose-candidates to accept a number instead of a string.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+                  "| Handle | Last update |"
+                  "|:-------|:------------|")
+            (map (match-lambda
+                   (#(handle last-update)
```
You don't need to use match twice.  You can untangle everything in the same match clause:

    (define results
        '(#(handle1 last-updated1)
          #(handle2 last-updated2)
          #(handle3 last-updated3)))

    (match results
      ((#(h l) ...)
        (pk 'hs h 'ls l)))

rekado replied:
This collates all handles in the list `h` and all updates in the list `l`.  That's because we tell `match` with the ellipses that `results` is uniform and repeating.

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+        (format #false "~a was deleted" handle))
+       (format #false "~a doesn't exist" handle)))
+    handles)
+   "\n"))
```
As a general pattern it may be better to write the messages to the `(current-output-port)` and not bother with `string-join`ing them.

The caller can then parameterize `current-output-port` to a string port if the messages should not go to stdout but to a string instead.

rekado replied:
Example:

    (define (noisy)
      (format #t "hello\n")
      (display "bye" (current-output-port))
      (display "world") (newline)
      (display "whatever") (newline))

    (call-with-output-string
      (lambda (string-port)
        (parameterize ((current-output-port string-port))
          (noisy))))

rekado replied:
This lets you focus on the important stuff and have the caller preserve output if it needs it.

jonasfreimuth committed:
subcommand-update: Fix parsing of n for choose

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+          "SELECT * FROM candidates WHERE handle = :handle"
+          #:handle handle)
+    (() #false)
+    (some #true)))
```
In Scheme predicates are often expected to go the extra mile.  `member`, for example, won't just tell you that an element is a member of a list, it will also return that member in the successful case.
Here the SQL query does some work: it collects the candidate fields (aside: use LIMIT to allow the SQL backend to stop early), but then you just throw them all away.  You could either return them all in the success case, or be explicit about the fact that you only care about whether or not a record exists by doing something like `SELECT 1 FROM candidates`.

rekado commented on mattermost.scm:
```diff
@@ -302,11 +310,16 @@ a colon-separated pair of hours and minutes."
+                #:whom whom
+                #:when (current-date (%tz))
+                #:recur? #false
+                #:message what))
```
is "remind <dude> now" essentially "tell <dude>"?

rekado commented on mattermost.scm:
```diff
@@ -383,36 +396,323 @@ a colon-separated pair of hours and minutes."
+         %db
+         (string-join 
+         '("DELETE FROM candidates"
+           "WHERE handle = :handle")
```
Here I'd prefer to filter the list of handles first, then issue *one* SQL statement for all deletions, and then return the messages with all deletions and all ignored handles.

You can use `partition` to partition `handles` by `handle-exists?`; it returns two values.  Use SRFI-71 to bind multiple values in a regular `let` expression:

    (use-modules (srfi srfi-71))
    (define (return-multiple-values)
      (values 42 "hello world"))

    (let ((existing ignored  (return-multiple-values)))
        (pk 'existing existing 'ignored ignored))

rekado reviewed (changes requested):
Excellent first Scheme contribution!

In addition to resolution of the comments, I'd like to see some tests and docstrings, especially for the choose procedures.

jonasfreimuth committed:
Fix message for update choose

  The tilde originally initiated a format ecape sequence for a tab char.

jonasfreimuth committed:
Add initialization of random state

  This is to ensure non-reproducibility, see
  <https://github.com/BIMSBbioinfo/sudo/pull/2#discussion_r1118483541>.

jonasfreimuth committed:
weighted-sample: Slightly clarify appending choice

  See
  <https://github.com/BIMSBbioinfo/sudo/pull/2#discussion_r1118485826>.

jonasfreimuth committed:
choose-candidates: Simplify weights creation

  See
  <https://github.com/BIMSBbioinfo/sudo/pull/2#discussion_r1118486930>.

jonasfreimuth committed:
choose-candidates: Simplify entry deconstruction

  See
  <https://github.com/BIMSBbioinfo/sudo/pull/2#discussion_r1118487739>.

jonasfreimuth committed:
handle-exists?: limit query to one entry

  See
  <https://github.com/BIMSBbioinfo/sudo/pull/2/files#r1118618818>.

jonasfreimuth committed:
handle-exists?: Add returning of entry

  See
  <https://github.com/BIMSBbioinfo/sudo/pull/2/files#r1118618818>.

jonasfreimuth committed:
choose-candidates: Move coercion of n out of proc

  See
  <https://github.com/BIMSBbioinfo/sudo/pull/2/files#r1118511642>.

jonasfreimuth committed:
weighted-sample: Rename value-weights var

  Rereading this it seems better than having two different `x`s.

jonasfreimuth committed:
weighted-sample: Handle duplicate x entries

  Previously, if x had duplicate entries with differing weights, the
  use of `assoc-remove!` could lead to the element with the wrong weight
  being removed.

jonasfreimuth committed:
weighted-sample: Fix indentation

jonasfreimuth committed:
weighted-sample: Add docstring

jonasfreimuth committed:
weighted-random: Handle case of all weights zero

  A zero weight leads to the probability of the pick being zero, which
  is a problem if all weights are zero, which here would lead to an
  indexing error. Although that would be the logically correct thing to
  do, in this context it is more useful to treat this case as special
  and treat the weights all equal and sample uniformly.

jonasfreimuth committed:
weighted-random*: Rename procedure

  Due to the potentially unexpected behaviour regarding all-zero
  weights, this seems appropriate.

jonasfreimuth committed:
weighted-random*: Add docstring

jonasfreimuth committed:
tests.scm: Add tests on random & sampling procs

  There is still some error handling to do, but first the procedurs
  would need error handling in the first place.
